File: cousinInBinaryTree.java
//Time Complexity: O(n)
//Space Complexity: O(h), h is the height of the tree for dfs for bfs O(n)
//Did this code successfully run on Leetcode : yes
//Any problem you faced while coding this : No
//bfs
class Solution {
    
    public boolean isCousins(TreeNode root, int x, int y) {
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            int size = q.size();
            boolean x_flag = false;
            boolean y_flag = false;
            for(int i = 0; i < size; i++){
                TreeNode curr = q.poll();
                if(curr.val == x) x_flag = true;
                if(curr.val == y) y_flag = true;
                if(curr.left != null && curr.right != null){
                    if(curr.left.val == x && curr.right.val == y) return false;
                    if(curr.left.val == y && curr.right.val == x) return false;
                }
                if(curr.left != null) q.add(curr.left);
                if(curr.right != null) q.add(curr.right);
            }
            if(x_flag && y_flag) return true;
            if(x_flag || y_flag) return false;
        }
        return false;
    }
    
}


/* DFS
class Solution {
    TreeNode x_parent;
    TreeNode y_parent;
    int x_ht;; int y_ht;
    public boolean isCousins(TreeNode root, int x, int y) {
        if(root == null) return false;
        dfs(root, null, 0, x,y);
        return x_parent != y_parent && x_ht == y_ht;
    }
    private void dfs(TreeNode root, TreeNode parent, int level, int x, int y){
        //base
        if(root == null) return;
        //logic
        if(root.val == x){
            x_parent = parent;
            x_ht = level;
        }
        if(root.val == y){
            y_parent = parent;
            y_ht = level;
        }
        if(x_parent == null || y_parent == null)
            dfs(root.left, root, level + 1, x, y);
        if(x_parent == null || y_parent == null)
            dfs(root.right, root, level + 1, x, y);
    }
}
*/

File: CousinInBinaryTree.py
class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        x_flag = False
        y_flag = False
        
        queue = collections.deque()
        
        if root == None:
            return False
        
        queue.append([root, None, 0])
        result = []
        
        while queue:
            
            node, parent, level = queue.popleft()
            
            # base condition
            
            if len(result) == 2:
                break
                
            if node.val == x or node.val == y:
                result.append((parent, level))
                
            if node.left:
                queue.append((node.left, node, level + 1))
            if node.right:
                queue.append((node.right, node, level + 1))
                
                
        node_x, node_y = result
        
        return node_x[0] != node_y[0] and node_x[1] == node_y[1]
            
            
    # BFS Approach

    # TIME COMPLEXITY: O(N) for visiting all the noes
    # SPACE COMPLEXITY: O(D) where D is the diameter of the tree. It will be N/2 in case of complete binary tree
            
class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        def dfs(root, parent, level):
            nonlocal x_level
            nonlocal x_parent
            nonlocal y_level
            nonlocal y_parent
            
            if root != None:
                if root.val == x:
                    x_level = level
                    x_parent = parent
                if root.val == y:
                    y_level = level
                    y_parent = parent
                    
                dfs(root.left, root, level + 1)
                dfs(root.right, root, level + 1)
        
        x_level = 0
        y_level = 0
        x_parent = TreeNode(-1)
        y_parent = TreeNode(-1)
        dfs(root, TreeNode(-1), 0)
        
        if x_level == y_level and x_parent != y_parent:
            return True
        else:
            return False
        
        # TIME COMPLEXITY: O(N) where n is the total number of nodes 
        # SPACE COMPLEXITY: O(H) where H is the height of the tree. O(n) in worst case